### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by shushaaniik**

### CODE REVIEW

File: inf_post/build/CMakeCache.txt [hunk 1]
```diff
@@ -0,0 +1,369 @@
+# This is the CMakeCache file.
```
Comment (matevosmehrabyan) [inline inf_post/build/CMakeCache.txt]: This file must not be in git

File: inf_post/build/CMakeFiles/3.22.1/CMakeCCompiler.cmake [hunk 2]
```diff
@@ -0,0 +1,72 @@
+set(CMAKE_C_COMPILER "/usr/bin/cc")
```
Comment (matevosmehrabyan) [inline inf_post/build/CMakeFiles/3.22.1/CMakeCCompiler.cmake]: remove build directory from git

File: inf_post/include/balanced.h [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#include <stack>
```
Comment (matevosmehrabyan) [inline inf_post/include/balanced.h]: Why have you written down each function in a separate file????? They all are checks. combine them

File: inf_post/include/infix.h [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#include <string>
+#include <stack>
```
Comment (matevosmehrabyan) [inline inf_post/include/infix.h]: You have included <stack> but it is not used in this file. don't do so.

File: inf_post/src/balanced.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+#include "balanced.h"
+
+bool balanced(std::string sequance)
+{
+    std::stack<char> s;
+    for (int i = 0; i < sequance.length(); i++) 
+    {
+	if(sequance[i] == '(' || sequance[i] == ')')
+	{
+            if (s.empty())
```
Comment (matevosmehrabyan) [inline inf_post/src/balanced.cpp]: Always use brackets

File: inf_post/src/infix.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#include <iostream>
+#include "infix.h"
+
+std::string infix(std::string postfix)
```
Comment (matevosmehrabyan) [inline inf_post/src/infix.cpp]: rename. from the name it is not clear what it does

File: inf_post/src/balanced.cpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#include <iostream>
+#include "balanced.h"
+
+bool balanced(std::string sequance)
```
Comment (matevosmehrabyan) [inline inf_post/src/balanced.cpp]: As Tigran said. "Functions which return boolean, their names must start with 'is' "

File: inf_post/src/op_priority.cpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
+#include <iostream>
+#include "op_priority.h"
+
+int op_priority(char c)
+{
+    switch(c)
+    {
+        case '+':
+        case '-':
+            return 1;
```
Comment (matevosmehrabyan) [inline inf_post/src/op_priority.cpp]: Create an enum for priorities

File: word_occupancy/src/map_create.cpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
+#include <iostream>
+#include "map_create.h"
+
+void map_create(std::string& s, std::unordered_map<std::string, int>& um)
```
Comment (matevosmehrabyan) [inline word_occupancy/src/map_create.cpp]: rename

File: word_occupancy/src/main.cpp [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+#include <unordered_map>
+#include "print.h"
+#include "map_create.h"
+
+int main()
+{
+    std::unordered_map<std::string, int> umap;
+    std::string input;
+
+    getline(std::cin, input);
```
Comment (matevosmehrabyan) [inline word_occupancy/src/main.cpp]: Upgrade your program. Now your program must get a file path and read the string from there

File: inf_post/src/infix.cpp [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+	    temp += op2;
+	    s.pop();
+	    temp += ')';
+	    s.push(temp);
+	}
+    }
+    result = s.top();
+    s.pop();
+    if(!s.empty())
+        return "Non valid input";
```
Comment (matevosmehrabyan) [inline inf_post/src/infix.cpp]: This is a bad idea. C++ has an Exception mechanism. use it.

File: inf_post/include/infix.h (hunk 4 above)
Comment (shushaaniik) [reply inf_post/include/infix.h]: +

File: inf_post/build/CMakeCache.txt (hunk 1 above)
Comment (shushaaniik) [reply inf_post/build/CMakeCache.txt]: +

File: inf_post/build/CMakeFiles/3.22.1/CMakeCCompiler.cmake (hunk 2 above)
Comment (shushaaniik) [reply inf_post/build/CMakeFiles/3.22.1/CMakeCCompiler.cmake]: +

File: inf_post/src/balanced.cpp (hunk 5 above)
Comment (shushaaniik) [reply inf_post/src/balanced.cpp]: +

File: inf_post/src/balanced.cpp (hunk 7 above)
Comment (shushaaniik) [reply inf_post/src/balanced.cpp]: +

File: inf_post/include/balanced.h (hunk 3 above)
Comment (shushaaniik) [reply inf_post/include/balanced.h]: +

File: inf_post/src/infix.cpp (hunk 6 above)
Comment (shushaaniik) [reply inf_post/src/infix.cpp]: +

File: word_occupancy/src/map_create.cpp (hunk 9 above)
Comment (shushaaniik) [reply word_occupancy/src/map_create.cpp]: +

File: inf_post/src/infix.cpp (hunk 11 above)
Comment (shushaaniik) [reply inf_post/src/infix.cpp]: +

File: word_occupancy/src/main.cpp (hunk 10 above)
Comment (shushaaniik) [reply word_occupancy/src/main.cpp]: +

File: inf_post/src/op_priority.cpp (hunk 8 above)
Comment (shushaaniik) [reply inf_post/src/op_priority.cpp]: +

File: inf_post/include/checks.h [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#include <string>
+
```
Comment (matevosmehrabyan) [inline inf_post/include/checks.h:5]: Where are #ifndef - #define -s ?

File: inf_post/include/op_priority.h [hunk 13]
```diff
@@ -0,0 +1,3 @@
+enum Operator_Priority { not_operator, first, second, third };
```
Comment (matevosmehrabyan) [inline inf_post/include/op_priority.h]: It's not clear. It turns out that Operator_Priority::first < Operator_Priority::third......

File: word_occupancy/src/main.cpp (hunk 10 above)
Comment (matevosmehrabyan) [reply word_occupancy/src/main.cpp]: File path must be passed as program argument, not STDIN

File: word_occupancy/src/main.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        if ( lstat(file_name.c_str(), &s) == 0 ) 
+        {
+            if ( S_ISDIR(s.st_mode) )
+	    {
+                throw(1);
+	    }
+        } 
+
+	std::ifstream input_file;
+	input_file.open(file_name);
```
Comment (matevosmehrabyan) [inline word_occupancy/src/main.cpp]: You have not closed the file

File: word_occupancy/src/main.cpp [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                std::getline (input_file, temp);
+		sentence += temp;
+            }
+            make_word_table(sentence, umap);
+            print(umap);
+	}
+    }
+    catch(std::string s) 
+    {
+        std::cout << "Couldn't open file " << s << "\n";
```
Comment (matevosmehrabyan) [inline word_occupancy/src/main.cpp]: return with error

File: word_occupancy/src/main.cpp [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    catch(int x)
+    {
+        std::cout << "The input is directory\n";
```
Comment (matevosmehrabyan) [inline word_occupancy/src/main.cpp]: return with error

File: word_occupancy/src/make_word_table.cpp [hunk 17]
```diff
@@ -0,0 +1,26 @@
+#include <iostream>
+#include "make_word_table.h"
+
+void make_word_table(std::string& s, std::unordered_map<std::string, int>& um)
```
Comment (matevosmehrabyan) [inline word_occupancy/src/make_word_table.cpp]: bad variable name 'um'. rename

File: inf_post/src/main.cpp [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+int main()
+{
+    try
+    {
+        std::string input;
+        getline(std::cin, input);
+        input = no_spaces(input);
+        if(!(is_valid(input)))
+        {
+            throw(input);
```
Comment (matevosmehrabyan) [inline inf_post/src/main.cpp]: The way you changed the code, now you don't need to throw exception. if it is not valid, just return with error

File: inf_post/src/checks.cpp [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        if(is_operand(s[i]))
+	{
+	    operand_count++;
+	}
+	if(is_operator(s[i]))
+	{
+	    operator_count++;
+	}
+    }
+    return (operand_count-operator_count == 1);
```
Comment (matevosmehrabyan) [inline inf_post/src/checks.cpp]: redunant parentses

File: inf_post/src/checks.cpp [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+	{
+            return false;
+	}
+    }
+    return true;
+}
+
+bool is_infix(std::string& s)
+{
+    return !is_operator(s[s.size()-1]);
```
Comment (matevosmehrabyan) [inline inf_post/src/checks.cpp]: string has '[back](https://en.cppreference.com/w/cpp/string/basic_string/back)' which allows to access the last element

File: inf_post/src/checks.cpp [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+bool is_operator(char c)
+{
+    return (c == '+' || c == '-' || c == '*' || c == '/' || c == '%');
+}
+
+bool is_operand(char c)
+{
+    return((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'));
```
Comment (matevosmehrabyan) [inline inf_post/src/checks.cpp]: Outer level's parentses are redunant

File: inf_post/include/checks.h (hunk 12 above)
Comment (shushaaniik) [reply inf_post/include/checks.h:5]: +

File: word_occupancy/src/main.cpp (hunk 10 above)
Comment (shushaaniik) [reply word_occupancy/src/main.cpp]: +

File: word_occupancy/src/main.cpp (hunk 15 above)
Comment (shushaaniik) [reply word_occupancy/src/main.cpp]: +

File: word_occupancy/src/main.cpp (hunk 16 above)
Comment (shushaaniik) [reply word_occupancy/src/main.cpp]: +

File: word_occupancy/src/make_word_table.cpp (hunk 17 above)
Comment (shushaaniik) [reply word_occupancy/src/make_word_table.cpp]: +

File: inf_post/src/checks.cpp (hunk 19 above)
Comment (shushaaniik) [reply inf_post/src/checks.cpp]: +

File: inf_post/src/checks.cpp (hunk 20 above)
Comment (shushaaniik) [reply inf_post/src/checks.cpp]: +

File: inf_post/src/checks.cpp (hunk 21 above)
Comment (shushaaniik) [reply inf_post/src/checks.cpp]: +

File: inf_post/src/main.cpp (hunk 18 above)
Comment (shushaaniik) [reply inf_post/src/main.cpp]: +

File: inf_post/include/op_priority.h (hunk 13 above)
Comment (shushaaniik) [reply inf_post/include/op_priority.h]: +

File: word_occupancy/src/main.cpp [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+#include <fstream>
+#include <unordered_map>
+#include <sys/stat.h>
+#include "print.h"
+#include "make_word_table.h"
+
+int main(int argc, char** argv)
+{
+    std::unordered_map<std::string, int> umap;
+    std::string file_name = argv[1];
```
Comment (matevosmehrabyan) [inline word_occupancy/src/main.cpp:18]: What if no argument was passed? it throws Abort. fix it

File: inf_post/include/checks.h [hunk 23]
```diff
@@ -0,0 +1,14 @@
+#ifndef CHECKS_H
+#define CHECKS_h
```
Comment (matevosmehrabyan) [inline inf_post/include/checks.h]: defined var and the upper check's  differ

File: inf_post/include/op_priority.h [hunk 24]
```diff
@@ -0,0 +1,8 @@
+#ifndef OP_PRIORITY_H
+#define OP_PRIORITY_H
+
+enum Operator_Priority { third, second, first, not_operator };
```
Comment (matevosmehrabyan) [inline inf_post/include/op_priority.h]: remove not_operator from enum

File: inf_post/src/op_priority.cpp [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+            break;
+        case '*':
+        case '/': 
+	    return second;
+	    break;
+        case '%': 
+            return third;
+            break;
+        default:
+            return not_operator;
```
Comment (matevosmehrabyan) [inline inf_post/src/op_priority.cpp]: throw an exception in this case

File: inf_post/src/op_priority.cpp [hunk 26]
```diff
@@ -0,0 +1,22 @@
+#include <iostream>
+#include "op_priority.h"
+
+Operator_Priority op_priority(char c)
```
Comment (matevosmehrabyan) [inline inf_post/src/op_priority.cpp]: rename variable 'c'

File: word_occupancy/src/main.cpp (hunk 14 above)
Comment (shushaaniik) [reply word_occupancy/src/main.cpp]: +

File: inf_post/include/checks.h (hunk 23 above)
Comment (shushaaniik) [reply inf_post/include/checks.h]: +

File: word_occupancy/src/main.cpp (hunk 22 above)
Comment (shushaaniik) [reply word_occupancy/src/main.cpp:18]: +

File: inf_post/include/op_priority.h (hunk 24 above)
Comment (shushaaniik) [reply inf_post/include/op_priority.h]: +

File: inf_post/src/op_priority.cpp (hunk 25 above)
Comment (shushaaniik) [reply inf_post/src/op_priority.cpp]: +

File: inf_post/src/op_priority.cpp (hunk 26 above)
Comment (shushaaniik) [reply inf_post/src/op_priority.cpp]: +

File: inf_post/src/main.cpp [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        if(is_infix(input))
+        {
+	    try
+	    {
+                std::cout << convert_to_postfix(input) << std::endl;
+	    }
+	    catch(char)
+	    {
+		std::cerr << "Wrong use of program\n";
```
Comment (matevosmehrabyan) [inline inf_post/src/main.cpp:29]: return 1

**Pull request merged**